## 01 Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

## 02 Importing Datasets

In [2]:
path = r'C:\Users\monik\Desktop\Career Foundry\Course\Data Immersion\4 Python\07.2025 Instacart Basket Analysis - Monika Kaniewska'

In [3]:
# Importing product dataset

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),index_col=False)

In [5]:
# Importing orders wrangled dataset

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

## 03 Consistency Checks df_prods

In [7]:
# creating dataframe
df_test = pd.DataFrame()

In [8]:
# creating a mixed type column
df_test['mix'] = ['a','b',1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Mixed-Type Data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\monik\AppData\Local\Temp\ipykernel_12600\3927803157.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [11]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


In [12]:
# Fixing mix type columns
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
df_test['mix'].head()

0       a
1       b
2       1
3    True
Name: mix, dtype: object

In [14]:
# Missing Values

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
# Addressing Missing Values

In [19]:
df_prods.shape

(49693, 5)

In [20]:
# Creating new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
# Another way of dropping missing values
df_prods.dropna(inplace = True)

In [23]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [24]:
# Finding Dublicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
# Addressing Duplicates

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

In [30]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [52]:
df_prods_clean_no_dups.shape

(49672, 5)

## 04 Consistency Checks df_ords

In [31]:
#2. Run the df.describe() function on your df_ords dataframe. 

In [32]:
df_ords.describe()

Unnamed: 0      order_id       user_id  purchase_sequence  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05       1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04       1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00       1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04       5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05       1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05       2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05       1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [33]:
# Dropping the Unnamed:0 column - index

In [34]:
df_ords = df_ords.drop(columns = ['Unnamed: 0']) 

In [35]:
df_ords.describe()

order_id       user_id  purchase_sequence  orders_day_of_week  \
count  3.421083e+06  3.421083e+06       3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05       1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04       1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00       1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04       5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05       1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05       2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05       1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [36]:
pd.set_option('display.float_format', '{:.2f}'.format) 

In [37]:
df_ords.describe()

order_id    user_id  purchase_sequence  orders_day_of_week  \
count 3421083.00 3421083.00         3421083.00          3421083.00   
mean  1710542.00  102978.21              17.15                2.78   
std    987581.74   59533.72              17.73                2.05   
min         1.00       1.00               1.00                0.00   
25%    855271.50   51394.00               5.00                1.00   
50%   1710542.00  102689.00              11.00                3.00   
75%   2565812.50  154385.00              23.00                5.00   
max   3421083.00  206209.00             100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

The count of days_since_prior_order is lower than in the other columns, which might indicate missing values—likely the first orders for each user.
The maximum value in purchase_sequence is 100, while the median is 11 and the standard deviation is 17.15. This suggests the maximum might be an outlier.
The order_day_of_week ranges from 0 to 6 (7 days total), which is correct. 
Similarly, order_hour_of_day ranges from 0 to 23 (24 hours), which is also correct.

In [38]:
#3. Check for mixed-type data in your df_ords dataframe.

In [39]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\monik\AppData\Local\Temp\ipykernel_12600\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\monik\AppData\Local\Temp\ipykernel_12600\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\monik\AppData\Local\Temp\ipykernel_12600\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\monik\AppData\Local\Temp\ipykernel_12600\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\monik\AppData\Local

In [40]:
#4.If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [41]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   purchase_sequence       int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


I haven't found mixed-type data.

In [42]:
# 5. Run a check for missing values in your df_ords dataframe.

In [43]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
purchase_sequence              0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [44]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [45]:
df_ords_nan

order_id  user_id  purchase_sequence  orders_day_of_week  \
0         2539329        1                  1                   2   
11        2168274        2                  1                   2   
26        1374495        3                  1                   1   
39        3343014        4                  1                   6   
45        2717275        5                  1                   3   
...           ...      ...                ...                 ...   
3420930    969311   206205                  1                   4   
3420934   3189322   206206                  1                   3   
3421002   2166133   206207                  1                   6   
3421019   2227043   206208                  1                   1   
3421069   3154581   206209                  1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

There are 206,209 missing values in the dataset. The column 'days_since_prior_order' contains missing data for each customer's first order (where the purchase_sequence is 1), since there was no previous order to compare to.

In [46]:
# 6. Address the missing values using an appropriate method.

Initially, I considered replacing the missing values in 'days_since_prior_order' with 0, since no days passed before the first order. However, doing so would have impacted the statistics. Since these missing values correspond to first-time orders where no previous order was made, they are not errors. Therefore, there is no need to remove or replace them. Because of this, I will keep them as they are.

In [47]:
# 7. Run a check for duplicate values in your df_ords data.
# 8. Address the duplicates using an appropriate method.

In [48]:
df_dups_ords = df_ords[df_ords.duplicated()]

In [49]:
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, purchase_sequence, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There were no duplicate values for found in df_ords. However, if there had been some duplicates, I would have deleted them.

In [50]:
# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [51]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))